File: examples/mathkit/verify/verify.go
// The Go driver of the FFI bench: constructs the generated client against the
// real stand-in library (no stub anywhere) and checks the values that come
// back through both operations. The fallback answers its first calculator,
// the constant one, so combined_value is the base; the series answers its
// last value.
package main

import (
	"context"
	"fmt"
	"os"

	"example.com/mathkit/mathkit"
)

func main() {
	c, err := mathkit.New(1.5, "2 * 3", 2, []float64{1, 2, 3})
	if err != nil {
		fail("construct: %v", err)
	}
	combined, err := c.CombinedValue(context.Background())
	if err != nil {
		fail("combined_value: %v", err)
	}
	if combined != 1.5 {
		fail("combined_value: want 1.5, got %v", combined)
	}
	series, err := c.SeriesValue(context.Background())
	if err != nil {
		fail("series_value: %v", err)
	}
	if series != 3 {
		fail("series_value: want 3, got %v", series)
	}
	fmt.Println("ffi bench (go): ok")
}

func fail(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

File: examples/mathkit/verify/verify.rs
//! The Rust driver of the FFI bench: constructs the generated client against
//! the real stand-in crate (no stub anywhere) and checks the values that come
//! back through both operations. The fallback answers its first calculator,
//! the constant one, so combined_value is the base; the series answers its
//! last value.
use mathkit_sdk::mathkit::Client;

#[tokio::main]
async fn main() {
    let client = Client::new(1.5, "2 * 3".to_string(), 2, vec![1.0, 2.0, 3.0])
        .await
        .expect("construct");
    let combined = client.combined_value().await.expect("combined_value");
    assert_eq!(combined, 1.5, "combined_value");
    let series = client.series_value().await.expect("series_value");
    assert_eq!(series, 3.0, "series_value");
    println!("ffi bench (rust): ok");
}
